libs/labelbox/src/labelbox/schema/workflow/workflow_utils.py

```diff
@@ -0,0 +1,408 @@
+                            "reason": "has no outgoing connections",
+                            "node_id": node.id,
+                            "node_type": node_type,
+                        }
+                    )
+
+        return errors
+
+    @classmethod
+    def validate(cls, workflow: "ProjectWorkflow") -> "ProjectWorkflow":
+        """Validate the workflow graph structure to identify potential issues."""
+        errors = []
+        nodes = workflow.get_nodes()
+        edges = workflow.get_edges()
+
+        if not nodes:
+            return workflow
+
+        # Build graph for validation
+        graph = ProjectWorkflowGraph()
+        for edge in edges:
+            graph.add_edge(edge.source, edge.target)
+
+        # Check for validation errors
+        initial_node_errors = cls.validate_initial_nodes(nodes)
+        errors.extend(initial_node_errors)
+
+        connection_errors = cls.validate_node_connections(nodes, graph)
+        errors.extend(connection_errors)
+
+        # Store validation results
+        workflow._validation_errors = {"validation": errors}
+        return workflow
+
+    @staticmethod
+    def check_validity(
+        workflow: "ProjectWorkflow",
+    ) -> Dict[str, List[Dict[str, str]]]:
+        """Check the validity of the workflow configuration."""
+        # Run validation
+        WorkflowValidator.validate(workflow)
+        # Return the validation errors
+        return WorkflowValidator.get_validation_errors(workflow)
+
+    @staticmethod
+    def get_validation_errors(
+        workflow: "ProjectWorkflow",
+    ) -> Dict[str, List[Dict[str, str]]]:
+        """Get validation errors from the most recent validation."""
+        if "errors" not in workflow._validation_errors:
+            # Run validation if not already done
+            WorkflowValidator.validate(workflow)
+        return workflow._validation_errors
+
+    @staticmethod
+    def format_validation_errors(
+        validation_errors: Dict[str, List[Dict[str, str]]],
+    ) -> str:
+        """Format validation errors into a human-readable string."""
+        errors = validation_errors.get("errors", [])
+        if not errors:
+            return ""
+
+        error_details = []
+        for error in errors:
+            node_id = error.get("node_id", "unknown")
+            node_type = error.get("node_type", "unknown")
+            reason = error.get("reason", "unknown reason")
+
+            # Extract additional details if available
+            details = error.get("details", "")
+            error_msg = f"Node {node_id[:8]}... ({node_type}) {reason}"
+            if details:
+                error_msg += f" - {details}"
+
+            error_details.append(error_msg)
+
+        return f"Workflow validation found the following issues: {'; '.join(error_details)}"
+
+
+class WorkflowLayoutManager:
+    """Layout management utilities for workflow visualization."""
+
+    @staticmethod
+    def reposition_nodes(
```

Comment:
It's unfortunate we have to include code that's concerned with the presentation layer in the SDK - I think I understand why based on our previous conversation but in the future let's have a conversation with stakeholders before going down this path.